op-service/sources/l1_beacon_client.go

```diff
@@ -0,0 +1,173 @@
+	timeToSlotFn TimeToSlotFn
+}
+
+// NewL1BeaconClient returns a client for making requests to an L1 consensus layer node.
+func NewL1BeaconClient(cl client.HTTP) *L1BeaconClient {
+	return &L1BeaconClient{cl: cl}
+}
+
+func (cl *L1BeaconClient) apiReq(ctx context.Context, dest any, method string) error {
+	headers := http.Header{}
+	headers.Add("Accept", "application/json")
+	resp, err := cl.cl.Get(ctx, method, headers)
+	if err != nil {
+		return fmt.Errorf("%w: http Get failed", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		errMsg, _ := io.ReadAll(resp.Body)
+		_ = resp.Body.Close()
+		return fmt.Errorf("failed request with status %d: %s", resp.StatusCode, string(errMsg))
+	}
+	if err := json.NewDecoder(resp.Body).Decode(dest); err != nil {
+		_ = resp.Body.Close()
+		return err
+	}
+	if err := resp.Body.Close(); err != nil {
+		return fmt.Errorf("%w: failed to close response body", err)
+	}
+	return nil
+}
+
+type TimeToSlotFn func(timestamp uint64) (uint64, error)
+
+// GetTimeToSlotFn returns a function that converts a timestamp to a slot number.
+func (cl *L1BeaconClient) GetTimeToSlotFn(ctx context.Context) (TimeToSlotFn, error) {
+	cl.initLock.Lock()
+	defer cl.initLock.Unlock()
+	if cl.timeToSlotFn != nil {
+		return cl.timeToSlotFn, nil
+	}
+
+	var genesisResp eth.APIGenesisResponse
+	if err := cl.apiReq(ctx, &genesisResp, genesisMethod); err != nil {
+		return nil, err
+	}
+
+	var configResp eth.APIConfigResponse
+	if err := cl.apiReq(ctx, &configResp, specMethod); err != nil {
+		return nil, err
+	}
+
+	genesisTime := uint64(genesisResp.Data.GenesisTime)
+	secondsPerSlot := uint64(configResp.Data.SecondsPerSlot)
+	if secondsPerSlot == 0 {
+		return nil, fmt.Errorf("got bad value for seconds per slot: %v", configResp.Data.SecondsPerSlot)
+	}
+	cl.timeToSlotFn = func(timestamp uint64) (uint64, error) {
+		if timestamp < genesisTime {
+			return 0, fmt.Errorf("provided timestamp (%v) precedes genesis time (%v)", timestamp, genesisTime)
+		}
+		return (timestamp - genesisTime) / secondsPerSlot, nil
+	}
+	return cl.timeToSlotFn, nil
+}
+
+// GetBlobSidecars fetches blob sidecars that were confirmed in the specified L1 block with the
+// given indexed hashes. Order of the returned sidecars is not guaranteed, and blob data is not
+// checked for validity.
+func (cl *L1BeaconClient) GetBlobSidecars(ctx context.Context, ref eth.L1BlockRef, hashes []eth.IndexedBlobHash) ([]*eth.BlobSidecar, error) {
+	if len(hashes) == 0 {
+		return []*eth.BlobSidecar{}, nil
+	}
+	slotFn, err := cl.GetTimeToSlotFn(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to get time to slot function", err)
+	}
+	slot, err := slotFn(ref.Time)
+	if err != nil {
+		return nil, fmt.Errorf("%w: error in converting ref.Time to slot", err)
+	}
+
+	builder := strings.Builder{}
+	builder.WriteString(sidecarsMethodPrefix)
+	builder.WriteString(strconv.FormatUint(slot, 10))
+	builder.WriteRune('?')
+	v := url.Values{}
+	for i := range hashes {
```

Comment:
Is there a limit on the number of hashes we can ask for in a single request?

Reply:
not that is documented, but it's going to be effectively limited by the max amount of blobs per block which is currently 6.

Reply:
Oh nice. I didn't know it was only 6 blobs per block. It feels very unlikely that you'd run into an issue with fetching 6 blobs in a single request.